drivers/hwmon_ina219.c

```diff
@@ -0,0 +1,339 @@
+
+		snprintf(ina219_base_path, sizeof(ina219_base_path),
+				"%s/%s", sysfs_hwmon_dir, entry->d_name);
+		closedir(sysfs);
+		return 0;
+	}
+
+	closedir(sysfs);
+	return -ENODEV;
+}
+
+static int update_intvar(
+	const char *base_path,
+	const char *name,
+	int *value)
+{
+	char path[PATH_MAX];
+	int ret;
+
+	if (snprintf(path, sizeof(path), "%s/%s", base_path, name) >= PATH_MAX) {
+		errno = ENAMETOOLONG;
+		upslog_with_errno(LOG_ERR, "snprintf(%s/%s) has failed", base_path, name);
+		return -ENAMETOOLONG;
+	}
+
+	if ((ret = file_read_number(path, value)) < 0) {
+		errno = -ret;
+		upslog_with_errno(LOG_ERR, "file_read_number(%s) has failed", path);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int update_voltage(void)
+{
+	return update_intvar(ina219_base_path, "in1_input", &voltage);
+}
+
+static int update_current(void)
+{
+	return update_intvar(ina219_base_path, "curr1_input", &current);
+}
+
+void upsdrv_makevartable(void)
+{
+	addvar(VAR_VALUE, "sysfs_dir",
+			"Path to sysfs dir of hwmon (" SYSFS_HWMON_DIR ")");
+	addvar(VAR_VALUE, "lb",
+			"Low battery threshold (" STRFY(LOW_BATTERY) ")");
+	addvar(VAR_VALUE, "hb",
+			"High battery threshold (" STRFY(HIGH_BATTERY) ")");
+}
+
+void upsdrv_initups(void)
+{
+	const char *hwmon_dir = SYSFS_HWMON_DIR;
+	int ret;
+
+	if (getval("sysfs_dir"))
+		hwmon_dir = getval("sysfs_dir");
+
+	if ((ret = detect_hwmon_ina219(hwmon_dir)) < 0) {
+		errno = -ret;
+		fatal_with_errno(EXIT_FAILURE, "detect_hwmon_ina219(%s) has failed",
+				hwmon_dir);
+	}
+
+	if (getval("lb")) {
+		const int tmp = atoi(getval("lb"));
+
+		if (tmp == 0 || tmp > 100) {
+			errno = ERANGE;
+			fatal_with_errno(EXIT_FAILURE, "Invalid value from option lb: '%s'",
+					getval("lb"));
+		}
+		else {
+			low_battery = tmp;
+		}
+	}
+
+	if (getval("hb")) {
+		const int tmp = atoi(getval("hb"));
+
+		if (tmp == 0 || tmp > 100) {
+			errno = ERANGE;
+			fatal_with_errno(EXIT_FAILURE, "Invalid value from option hb: '%s'",
+					getval("hb"));
+		}
+		else {
+			high_battery = tmp;
+		}
+	}
+
+	if (low_battery > high_battery) {
+		errno = EINVAL;
+		fatal_with_errno(EXIT_FAILURE, "Invalid options: lb > hb");
+	}
+
+	upslogx(LOG_NOTICE, "lb = %u hb = %u\n", low_battery, high_battery);
+}
+
+void upsdrv_initinfo(void)
+{
+	dstate_setinfo("ups.mfr", "%s", "Texas Instruments");
+	dstate_setinfo("ups.model", "%s", "INA219");
+	dstate_setinfo("ups.type", "%s", "ups");
+	dstate_setinfo("device.mfr", "%s", "Texas Instruments");
+	dstate_setinfo("device.model", "%s", "INA219");
+	dstate_setinfo("device.type", "%s", "ups");
+	dstate_setinfo("device.description", "%s",
+			"Bidirectional Current/Power Monitor With I2C Interface");
+
+	dstate_setinfo("battery.charge.low", "%d", low_battery);
+	dstate_setinfo("battery.charge.high", "%d", high_battery);
+}
+
+void upsdrv_updateinfo(void)
+{
+	unsigned int charge = 0;
+	int stale = 0;
+
+	if (update_voltage() < 0)
+		stale = 1;
+
+	upsdebugx(1, "Battery voltage: %.3fV", voltage / 1000.0);
```

Comment:
Consider a quieter debug level for the regular updates please. A single `-D` is not expected to be too noisy :)

Reply:
I put _3_ there.